### PR TITLE
Enhanced Database Reporting for MongoDB

### PR DESCRIPTION
**Pull request by romil-punetha**
## Which problem is this PR solving?

- Fixes [#1044](https://github.com/open-telemetry/opentelemetry-js/issues/1044)

## Short description of the changes

- The `enhancedDatabaseReporting` variable is used to allow capturing entire database queries instead of obfuscating query parameteres

### CODE REVIEW

**Comment by dyladan** [comment]
I renamed the PR because "0.7.1 proposal" doesn't really make sense.

**Comment by dyladan** [comment]
Why did you change the image used for mongo in tests?

**Comment by romil-punetha** [comment]
> Why did you change the image used for mongo in tests?

Have reverted that. By mistake.

File: examples/mongodb/client.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+function makeRequest() {
+  // span corresponds to outgoing requests. Here, we have manually created
+  // the span, which is created to track work that happens outside of the
+  // request lifecycle entirely.
+  const span = tracer.startSpan('makeRequest');
+
+  let queries = 0;
+  let responses = 0;
+
+  tracer.withSpan(span, () => {
```
Comment (dyladan) [inline examples/mongodb/client.js:17]: Both of these calls can be done in the same `withSpan` call.

File: plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts [hunk 2]
```diff
@@ -192,13 +192,17 @@ export class MongoDBPlugin extends BasePlugin<typeof mongodb> {
...
     if (command === undefined) return;
-    const query = Object.keys(command.query ?? command.q ?? command).reduce(
-      (obj, key) => {
-        obj[key] = '?';
-        return obj;
-      },
-      {} as { [key: string]: string }
-    );
+
+    const query = this._config?.enhancedDatabaseReporting
```
Comment (dyladan) [inline plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts:196]: I think we should add a comment to explain why this is done

**Comment by romil-punetha** [comment]
> Please add some documentation in the README that makes it clear what the enhanced db reporting flag does, and how to enable it.
> 
> I think the example can be drastically simplified. It doesn't need a client and server component, or to make http calls. A script that simply sets up the tracer and makes a call to mongodb should be enough to demonstrate the plugin.

I used the same format as mysql plugin.

File: plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts (hunk 2 above)
Comment (romil-punetha) [reply plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts:197]: done

**Comment by dyladan** [comment]
> > Please add some documentation in the README that makes it clear what the enhanced db reporting flag does, and how to enable it.
> > I think the example can be drastically simplified. It doesn't need a client and server component, or to make http calls. A script that simply sets up the tracer and makes a call to mongodb should be enough to demonstrate the plugin.
> 
> I used the same format as mysql plugin.

Yea I see that. We've been making an effort to simplify the examples recently though as we have too many that show too many things. Smaller examples are easier for beginners to understand.

**Comment by romil-punetha** [comment]
> i think we should document the option inside the readme of the plugin (like done with the http one [there](https://github.com/open-telemetry/opentelemetry-js/tree/master/packages/opentelemetry-plugin-http#http-plugin-options))

I dont see a folder for mongo plugin there.

**Comment by vmarchaud** [comment]
@romil-punetha I was refering to the documentation made **for** the http plugin options that you could re-use to document the `enhancedDatabaseReporting` inside the mongodb plugin's readme :)

**Comment by romil-punetha** [comment]
Oh, ok.

File: examples/mongodb/README.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
+# Overview
+
+OpenTelemetry Mongodb Instrumentation allows the user to automatically collect trace data and export them to the backend of choice (we can use Zipkin or Jaeger for this example), to give observability to distributed systems.
+
+This is a modification of the HTTP example that executes multiple parallel requests that interact with a Mongodb server backend using the `mongo` npm module. The example displays traces using multiple connection methods.
+- Create Collection Query
+- Insert Document Query
+- Fetch All Documents Query
+
+### Http Plugin Options
```
Comment (vmarchaud) [inline examples/mongodb/README.md:10]: This should be inside the plugin readme ([there](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/master/plugins/node/opentelemetry-plugin-mongodb/README.md)), not the example. Also the option of the plugin only contains `enhancedDatabaseReporting` with your PR, why did you add the others ?

Comment (romil-punetha) [reply examples/mongodb/README.md:10]: Moving this. 
I thought it would be good to have all available options documented in one place, and because the plugin takes all of these as input.

Comment (romil-punetha) [reply examples/mongodb/README.md:10]: @vmarchaud should i only add `enhancedDatabaseReporting` ?

Comment (vmarchaud) [reply examples/mongodb/README.md:10]: @romil-punetha yeah, we generally document only the custom options.

Comment (romil-punetha) [reply examples/mongodb/README.md]: Ok. Editing that.

**Comment by romil-punetha** [comment]
> lgtm, as @dyladan said the example could be simpler but thats not a blocker for me as generally user want examples that looks close to their use-cases which would be in this case a http server that make operation against mongodb.

Agreed. The same worked for me as well. The client-server example contains the minimum changes required instrument a server.

File: plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts [hunk 4]
```diff
@@ -192,13 +192,18 @@ export class MongoDBPlugin extends BasePlugin<typeof mongodb> {
...
-      (obj, key) => {
-        obj[key] = '?';
-        return obj;
-      },
-      {} as { [key: string]: string }
-    );
+
+    // capture parameters within the query as well if enhancedDatabaseReporting is enabled.
+    const query = this._config?.enhancedDatabaseReporting
+      ? command.query ?? command.q ?? command
```
Comment (obecny) [inline plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts:198]: Nit: line 198 and 199 has the same maybe consider this to put into some variable and then reuse ?
`const commandObj = command.query ?? command.q ?? command;`

File: plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts [hunk 5]
```diff
@@ -192,13 +192,18 @@ export class MongoDBPlugin extends BasePlugin<typeof mongodb> {
...
-    const query = Object.keys(command.query ?? command.q ?? command).reduce(
-      (obj, key) => {
-        obj[key] = '?';
-        return obj;
-      },
-      {} as { [key: string]: string }
-    );
+
+    // capture parameters within the query as well if enhancedDatabaseReporting is enabled.
+    const query = this._config?.enhancedDatabaseReporting
```
Comment (obecny) [inline plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts:197]: can you compare this directly with boolean expression
`this._config?.enhancedDatabaseReporting === true`
simply to avoid situation like this `enhancedDatabaseReporting = 'true'` or `enhancedDatabaseReporting='foo'` etc.

File: plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts (hunk 4 above)
Comment (romil-punetha) [reply plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts]: done

File: plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts (hunk 5 above)
Comment (romil-punetha) [reply plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts]: fixed this.
